pkg/runtime/core/trainingruntime.go

```diff
@@ -98,31 +105,86 @@ func (r *TrainingRuntime) buildObjects(
 		// The JobSetTemplateSpec annotations are overridden by the TrainJob Annotations (.spec.annotations).
 		propagationAnnotations[k] = v
 	}
+	jobSetSpecApply, err := apply.FromTypedObjWithFields[jobsetv1alpha2ac.JobSetSpecApplyConfiguration](&jobsetv1alpha2.JobSet{
```

Comment:
Ideally, to be strictly correct, the TrainingRuntime would be Get as unstructured in NewObjects to avoid pollution from zero / nil fields from the typed struct. It is converted to unstructured anyways.

Reply:
Yes, distinguishing zero /null is important for SSA.
So, Ideally, we want to get unstructured objects in https://github.com/kubeflow/trainer/blob/b89ce8491ff81b8b01513e26d0a917808b6eed89/pkg/runtime/core/trainingruntime.go#L75-L79. 

However, it will bring MPI unrelated change in this PR. So, let me open issue and do it as a follow-up, thanks.

Reply:
Opened: https://github.com/kubeflow/trainer/issues/2515